Implementing NumPy-like function torch.broadcast_to

RockingJavaBean opened the pull request:
Related #38349

Implement NumPy-like function `torch.broadcast_to` to broadcast the input tensor to a new shape.

RockingJavaBean committed:
implement torch.broadcast_to

RockingJavaBean commented:
This PR adds `torch.broadcast_to` function similar to [NumPy's broadcast_to](https://numpy.org/devdocs/reference/generated/numpy.broadcast_to.html), and its implementation is based on the existing `expand` operator.

The differences between `torch.broadcast_to` and the `expand` operator are below:
- `torch.broadcast_to` is implemented as both a torch function and a tensor method, while the `expand` is a tensor method.
- The signature of the `expand` operator is `expand(Tensor(a) self, int[] size, *, bool implicit=False)`, its `implicit` parameter is to let the `tracer` distinguish between expands inserted by broadcasts and those requested by the user.
  This parameter is not exposed by `torch.broadcast_to`.

Please kindly help review this PR. @mruberry

mruberry commented on torch/_tensor_docs.py:
```diff
@@ -769,6 +769,13 @@ def add_docstr_all(method, docstr):
 
+add_docstr_all('broadcast_to',
+               r"""
+broadcast_to(input, size) -> Tensor
```
"broadcast_to(input, size)" -> "broadcast_to(shape)"

The input is implied for methods, and the word "shape" is consistent with NumPy and other PyTorch functions.

RockingJavaBean replied:
Thanks so much for pointing this out, the method signature is updated accordingly in the latest PR.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1130,6 +1130,31 @@ def merge_dicts(*dicts):
 
+add_docstr(torch.broadcast_to,
+           r"""
+broadcast_to(input, size) -> Tensor
```
"size"->"shape"

RockingJavaBean replied:
Addressed.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1130,6 +1130,31 @@ def merge_dicts(*dicts):
+           r"""
+broadcast_to(input, size) -> Tensor
+
+Broadcasts the input tensor to a new shape according to :ref:`broadcasting-semantics`.
```
"Broadcasts :attr:\`input\` to the shape :shape:\`shape`. Equivalent to calling \`input.expand(shape)\`. See :meth:\`~Tensor.expand\` for details.

RockingJavaBean replied:
Thanks so much for this suggestion, the function description has been updated.
[rendered doc for torch.broadcast_to](https://user-images.githubusercontent.com/40605299/101917736-86db0080-3c03-11eb-9409-016d9b9c03c6.png)

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1130,6 +1130,31 @@ def merge_dicts(*dicts):
+
+Broadcasts the input tensor to a new shape according to :ref:`broadcasting-semantics`.
+
+.. warning::
```
This warning can be removed once expand is cited.

RockingJavaBean replied:
Addressed.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1130,6 +1130,31 @@ def merge_dicts(*dicts):
+
+Args:
+    {input}
+    size (tuple): a tuple defining the shape of the output tensor.
```
"(tuple)" -> "(list, tuple, or :class:\`torch.Size\`)"

RockingJavaBean replied:
Addressed.

mruberry commented on test/test_view_ops.py:
```diff
@@ -1038,6 +1038,34 @@ def test_broadcast_shapes(self, device):
 
+    # Skip BFloat16 since numpy does not support it
+    @dtypes(*torch.testing.get_all_dtypes(include_bfloat16=False))
+    def test_broadcast_to(self, device, dtype):
```
There's one quirk of torch.Tensor.expand that's different from NumPy's broadcast_to. NumPy's broadcast_to won't accept -1 as a value, but PyTorch's expand will.

NumPy:
```
a = np.array((1, 2, 3))
np.broadcast_to(a, (-1,))
: ValueError: all elements of broadcast shape must be non-negative
```

PyTorch:
```
t = torch.tensor((1, 2, 3))
t.expand(-1)
: tensor([1, 2, 3])
```

I think this discrepancy is OK, however, as this just adds functionality to NumPy's broadcast_to.

RockingJavaBean replied:
Thanks for the detailed illustration regarding the discrepancy between NumPy's broadcast_to and `torch.broadcast_to` function.
The check for `non-negative` shape is not added for `torch.broadcast_to`, making it behave consistently with `torch.Tensor.expand`.

mruberry commented on test/test_view_ops.py:
```diff
@@ -1038,6 +1038,34 @@ def test_broadcast_shapes(self, device):
+    # Skip BFloat16 since numpy does not support it
+    @dtypes(*torch.testing.get_all_dtypes(include_bfloat16=False))
+    def test_broadcast_to(self, device, dtype):
+        def can_broadcast(s0, s1):
```
It's great this this accounts for broadcast_to increasing the dimensionality of the tensor.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1109,6 +1109,13 @@ def method_tests():
         ('view_as', (S, S, S), (non_differentiable(torch.rand(S * S, S)),)),
         ('view_as', (), (non_differentiable(torch.tensor(5.5)),), 'scalar'),
         ('view_as', (), (non_differentiable(torch.rand(1, 1)),), 'scalar_to_dims'),
+        ('broadcast_to', (S, 1, 1), (dont_convert((S, S, S)),), '', (False,)),
```
We're actually starting to deprecate `method_tests()` entries like these in favor of writing OpInfos for new operators, like this:

https://github.com/pytorch/pytorch/blob/9417e92722f1b6f75e162cbe661d3ed788d8d37a/torch/testing/_internal/common_methods_invocations.py#L420

Basically, an OpInfo is added to this alphabetical list. It has to specify which dtypes it supports using the dtype functions defined here:

https://github.com/pytorch/pytorch/blob/9417e92722f1b6f75e162cbe661d3ed788d8d37a/torch/testing/__init__.py#L278

and provide the same information as these `method_tests()` entries like addmm does. For example, instead of these tuples the operator create a `sample_inputs_func` that returns them.

This is a new thing, which is a pain, but I think it will be helpful to learn and I would appreciate your feedback on it. It may be a little trickier to setup than `method_tests()` but it will make testing much easier.

One problem with OpInfos currently is that operations which support the bfloat16 datatype have to skip the following test:

```
skips=(
  # RuntimeError: "rsqrt_cuda" not implemented for 'BFloat16'
  SkipInfo(
    'TestCommon', 'test_variant_consistency_jit',
    device_type='cuda', dtypes=[torch.bfloat16])))
```

This is a bug in the test's implementation that we expect to fix soon. Please let me know if you have any questions about doing this. If creating an OpInfo is too much of a pain then we can ignore it for now and continue to use `method_tests`.

RockingJavaBean replied:
I'm truly grateful for this comprehensive explanation of the new `OpInfo` test framework.
I think it is much more descriptive and readable compared to the `method_tests`, as the long tuple line full of testing parameters is often hard to read.
Moreover, `OpInfo` is awesome because it provides the standard way of testing newly-added operators. It helps us develop code in a more TDD style and write more robust code.

RockingJavaBean replied:
I have rewritten the test using `OpInfo`, and updated the `test_ops.py` as `OpInfo` currently does not support testing operators whose parameters are not `torch.Tensor`.

mruberry replied:
I think it does support non-tensor parameters, but let me take a look and I'll comment inline.

mruberry reviewed (commented):
Hey @RockingJavaBean, thanks for implementing `broadcast_to`!

Overall this PR looks very good. I made a few suggestions in the documentation, and I'd like to suggest trying to replace the `method_tests` entries, which are being deprecated, with an OpInfo. Looking forward to hearing your thoughts!

RockingJavaBean committed:
update to OpInfo tests and improve docs

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into torch_broadcast_to

RockingJavaBean committed:
fix flake issue

RockingJavaBean commented:
@mruberry Thank you so much for the kind review and patient guidance on how to use `OpInfo` for testing, I have updated this PR with docs improvement and new `OpInfo` based test codes.

As the current implementation in `test_ops.py` uses the `clone` method for inputs, it would fail when the input is  `a sequence of ints` like `torch.broadcast_to`, or `a sequence of tensors` like PyTorch stack related functions.
https://github.com/pytorch/pytorch/blob/5ab90b2fda9a94e45c26be6e567e8d3fbf0e6dfa/test/test_ops.py#L208

This PR adds a helper method in `torch/testing/_internal/common_utils.py` to handle cloning for the above scenarios.
The `test/test_op_aliases.py`, which tests `torch.row_stack` as the alias of `torch.vstack`, is updated to use this common helper.

Please kindly take a look.

RockingJavaBean committed:
update clone_input_helper

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into torch_broadcast_to

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into torch_broadcast_to

mruberry commented on test/test_op_aliases.py:
```diff
@@ -201,7 +194,7 @@ def _fn(t):
                 script = fn_template.format(alias_name=info.alias_name, args=arg_string)
             else:
-                is_input_tensor_list = isinstance(info.get_input(device), collections.Sequence)
+                is_input_tensor_list = isinstance(info.get_input(device), Sequence)
```
I don't think you need to update this file, since this file doesn't currently use any OpInfos to test.

RockingJavaBean replied:
`Sequence` here is imported from `collections.abc` to replace `collections.Sequence`.
The reason for the change is that `collections.Sequence` will be deprecated after Python 3.9.
```python
In [4]: collections.Sequence
/root/anaconda3/envs/xx/bin/ipython:1: DeprecationWarning: Using or importing the ABCs from 'collections' 
instead of from 'collections.abc' is deprecated since Python 3.3,and in 3.9 it will stop working
```
Other changes for `test_op_aliases.py` has been reverted in the latest PR.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1130,6 +1130,26 @@ def merge_dicts(*dicts):
+           r"""
+broadcast_to(input, shape) -> Tensor
+
+Broadcasts :attr:`input` to the :attr:`shape`.
```
"to the :attr:\`shape\`" -> to the shape :attr:\`\shape\`"

Sorry this didn't format correctly the first time I suggested it. I think the formatting is fixed now.

RockingJavaBean replied:
Thanks so much for the kind suggestions for improving the doc, it has been updated accordingly.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -281,6 +281,22 @@ def sample_inputs_addmm(op_info, device, dtype, requires_grad):
+        ((), (1, 3, 2)),
+    )
+
+    return [SampleInput((make_tensor(size, device, dtype,
```
Return these as a tuple, not a list, because tuples in Python are immutable. Lists are appropriate only when a mutable iterable is needed.

RockingJavaBean replied:
Thank you so much for pointing this out, I updated the `sample_inputs_broadcast_to` to return a `tuple`.

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1804,6 +1805,18 @@ def get_int64_dtype(dtype):
+def clone_input_helper(input):
+    if isinstance(input, Sequence):
+        # use torch.clone for tensor, copy.deepcopy otherwise
+        return list(map(lambda x: torch.clone(x) if isinstance(x, torch.Tensor) else deepcopy(x),
```
Just map clone_input_helper so you're calling this recursively on sequences (that would also handle nested sequences)

RockingJavaBean replied:
Sorry for not considering the function of this helper carefully, the helper has been updated as a recursive method to support `nested sequences`.

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1804,6 +1805,18 @@ def get_int64_dtype(dtype):
+    if isinstance(input, torch.Tensor):
+        return torch.clone(input)
+    else:
+        return deepcopy(input)
```
I think we can get away without copying non-tensor inputs, since I don't think we have any functions that edit non-tensor inputs? 

Ideally we'd also only clone when we're performing the operation inplace, but that can be a TODO for a later PR.

RockingJavaBean replied:
Thanks for the advice, the `clone_input_helper` do not copy the non-tensor inputs in the latest PR.

RockingJavaBean replied:
`test_ops.py` is updated to only clone tensor-type inputs for inplace operations.
```python
variant_forward = variant(*(clone_input_helper(input) if variant is inplace else input
                            for input in sample.input),
                          *sample.args,
                          **sample.kwargs)
self.assertEqual(variant_forward, expected_forward)
```

mruberry reviewed (commented):
Hey @RockingJavaBean!

I really like the approach you took here. The new helper function for making copies makes sense. I have a few very small comments but also suggest reverting the changes to test_op_aliases.py, since those tests are not affected by this PR. In the future test_op_aliases.py will be refactored to use OpInfos, and that will be a better opportunity to update its test.

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into torch_broadcast_to

RockingJavaBean committed:
fix doc and tests

RockingJavaBean commented:
@mruberry I'm really thankful for your kind and thorough review. 

- the doc description of `torch.broadcast_to` has been updated.
- the changes besides the one for `collections.abc.Sequence` has been reverted for `test_op_aliases.py`.
- the new clone helper is changed to a recursive method so that nested sequences can be handled properly.

Please kindly take a look.

RockingJavaBean committed:
only copy tensor-type inputs for inplace operation

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into torch_broadcast_to

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into torch_broadcast_to

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -283,6 +283,22 @@ def sample_inputs_addmm(op_info, device, dtype, requires_grad):
                                     requires_grad=False))),)
 
+def sample_inputs_broadcast_to(op_info, device, dtype, requires_grad):
+    test_cases = (
```
This is a nice method of test generation.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -587,6 +603,16 @@ def sample_inputs_pinverse(op_info, device, dtype, requires_grad=False):
+           test_inplace_grad=False,
+           sample_inputs_func=sample_inputs_broadcast_to,
+           skips=(
+               # RuntimeError: "isfinite" not implemented for 'BFloat16'
```
After a rebase this skip should no longer be necessary - would you try removing it?

RockingJavaBean replied:
Thanks so much for pointing this out, `skips` has been removed and the Opinfo-based `broacast_to` tests are passed.

mruberry reviewed (commented):
Hey @RockingJavaBean, this looks great! Would you try removing the one skipped test and rebasing to resolve the merge conflict? After the rebase this should be OK to merge.

Looking forward to having `broadcast_to` in PyTorch!

RockingJavaBean committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into torch_broadcast_to

RockingJavaBean committed:
remove skipped tests

RockingJavaBean commented:
@mruberry I'm deeply grateful for the code review of this PR, and your exhaustive guidance regarding writing tests using the newly introduced `Opinfo`. 🎄 🎅 🎄
I have rebased the code and removed the skipped tests.

RockingJavaBean commented:
By clicking the [details pending CircleCI check](https://app.circleci.com/pipelines/github/pytorch/pytorch/253708/workflows/551f15fc-5c6f-4479-9938-3cab00a39007/jobs/9772548), all steps of `pytorch_linux_xenial_cuda10_2_cudnn7_py3_gcc7_build` are successful.
It seems the pending check is already completed while its status hasn't been synced to this PR page.

mruberry reviewed (approved):
Nice work, @RockingJavaBean!
